### ISSUE TITLE
s390x/NetworkManager: Connections do not work with inconsistent names e.g. eth1

### ISSUE THREAD

**Issue by jschintag** [coreos/fedora-coreos-config#1499]
When running the current ext.config.networking.* tests i see that the connections in the butane configs are picked up by Network Manager, but they are not linked to devices. This means that when the config calls for `eth1`, it creates the connection `eth1`, but then it goes on and creates `enc4` as a device, resulting in `ip addr show` showing `enc4`, instead of as expected `eth1`.

**Comment by miabbott** [comment]
@HuijingHei since you worked on some of these tests, could you have a look at how they work on the multi-arch platforms?

**Comment by HuijingHei** [comment]
Sure, I will try, but I do not have s390x platform, I will borrow one from beaker

**Comment by HuijingHei** [comment]
On s390x platform, run with CoreOS 410.84.202202090402-0 using `cosa kola qemuexec --qemu-image test.qcow2 --kargs net.ifnames=0 -c`, seems some kernel args are missed in the first boot

- After first booted and login, check `/proc/cmdline` kernel arg does not include `net.ifnames=0`, get `enc2` device

```
[core@cosa-devsh ~]$ cat /proc/cmdline | more
random.trust_cpu=on ignition.platform.id=qemu $ignition_firstboot ostree=/ostree/boot.1/rhcos/16a78463454c349f8a44c08964f1500e808c0dec8f8af24bf4fd2b0c0d58e583/0 ignition.firstboot
[core@cosa-devsh ~]$ rpm-ostree kargs list --deploy-index=0       
random.trust_cpu=on ignition.platform.id=qemu $ignition_firstboot ostree=/ostree/boot.1/rhcos/16a78463454c349f8a44c08964f1500e808c0dec8f8af24bf4fd2b0c0d58e583/0 net.ifnames=0 root=UUID=e17848c2-98a2-4779-bcf1-6c8927a9f5ef rw rootflags=prjquota boot=UUID=d3d02f95-d665-48dd-9689-1ff7881948ad
```

- Exec `systemctl reboot`, and check `/proc/cmdline` kernel arg include `net.ifnames=0`, get `eth0` device

```
[core@cosa-devsh ~]$ cat /proc/cmdline 
random.trust_cpu=on ignition.platform.id=qemu $ignition_firstboot ostree=/ostree/boot.1/rhcos/16a78463454c349f8a44c08964f1500e808c0dec8f8af24bf4fd2b0c0d58e583/0 net.ifnames=0 root=UUID=e17848c2-98a2-4779-bcf1-6c8927a9f5ef rw rootflags=prjquota boot=UUID=d3d02f95-d665-48dd-9689-1ff7881948ad
```

**Comment by HuijingHei** [comment]
Do more research, when running `cosa kola qemuexec --qemu-image test.qcow2 --kargs rd.break -c`, for the first boot, kernel args missed `rd.break` in the beginning and during initramfs, can not enter emergency mode. 
The karg is saved by `coreos-boot-edit`, after reboot, enter emergency mode

first boot logs:
`[    0.244808] Kernel command line: random.trust_cpu=on ignition.platform.id=qemu $ignition_firstboot ostree=/ostree/boot.1/rhcos/16a78463454c349f8a44c08964f1500e808c0dec8f8af24bf4fd2b0c0d58e583/0 ignition.firstboot`

after enter initramfs:
```
[    3.908512] systemd[1]: Started Apply Kernel Variables.
[    3.965117] dracut-cmdline[284]: dracut-410.84.202202090402-0 dracut-049-135.git20210121.el8
[    3.990043] dracut-cmdline[284]: Using kernel command line parameters: ip=auto rd.driver.pre=dm_multipath random.trust_cpu=on ignition.platform.id=qemu $ignition_firstboot ostree=/ostree/boot.1/rhcos/16a78463454c349f8a44c08964f1500e808c0dec8f8af24bf4fd2b0c0d58e583/0 ignition.firstboot
...
[   22.696235] coreos-boot-edit[909]: Injected kernel arguments into BLS: root=UUID=1cf6df6a-b4ea-4b5d-96da-412f3c5688ac rw rootflags=prjquota
[   22.773395] coreos-boot-edit[909]: random.trust_cpu=on ignition.platform.id=qemu $ignition_firstboot ostree=/ostree/boot.1/rhcos/16a78463454c349f8a44c08964f1500e808c0dec8f8af24bf4fd2b0c0d58e583/0 rd.break root=UUID=1cf6df6a-b4ea-4b5d-96da-412f3c5688ac rw rootflags=prjquota boot=UUID=5db0dc18-d151-41b5-8c1e-7dbadae1f7da
```

**Comment by dustymabe** [comment]
hmm. I don't think s390x uses grub does it? Maybe our `$ignition_firstboot` replacement code that happens on firstboot isn't working properly.

I wonder how this is handled for s390x in RHCOS.

**Comment by dustymabe** [comment]
It's also possible our `--kargs` implementation doesn't work correctly on `s390x`.

**Comment by jschintag** [comment]
As far as i remember the replacement code is only run during initramfs, therefore there is no way that the kargs are applied furing firstboot, but instead only after reboot.

**Comment by jschintag** [comment]
Here, to change kargs on s390x you need to call zipl, which happens here https://github.com/coreos/coreos-installer/blob/fb8f66071a300e11dbcf412d128bf2850464876d/src/bin/rdcore/kargs.rs#L51

**Comment by jschintag** [comment]
So i just tested it on s390x, and i can confirm that both `--kargs` and `--firstbootkargs` does not work on s390x, as it does not call `zipl`. So after calling `cosa run --kargs "foo"` i can see `foo` in `/boot/loader/entries/ostree-1-fedora-coreos.conf`. However `zipl` does not read that file on boot, so it is rather pointless.

What should work is to provide the tests with it's own ignition, like the kargs test does. I just tested that with the `ext.config.networking.kargs-rd-net` test, which suffers from the same problem. When you remove the appendKargs argument from the test and instead use an ignition file for the kargs, you use the kargs.service (which works).

I also just tested this with the `ext.config.networking.mtu-on-bond-ignition` test

**Comment by dustymabe** [comment]
> So i just tested it on s390x, and i can confirm that both `--kargs` and `--firstbootkargs` does not work on s390x, as it does not call `zipl`. So after calling `cosa run --kargs "foo"` i can see `foo` in `/boot/loader/entries/ostree-1-fedora-coreos.conf`. However `zipl` does not read that file on boot, so it is rather pointless.

I understand that `--kargs` edits the `/boot/loader/entries/ostree-1-fedora-coreos.conf` directly and won't work. However, IIUC, `--firstbootkargs` should write into `/boot/ignition.firstboot`. Does s390x platform not pick up arguments from that file like on others?

**Comment by jschintag** [comment]
I don't know what exactly happens there, but i do know that it ends up failling. My guess here would be that either it does not call `zipl` or it does not reboot after calling `zipl`.

**Comment by jlebon** [comment]
> My guess here would be that either it does not call `zipl`.

Yup, this is correct. AFAICT, there's nothing in kola that causes `zipl` to be run when modifying either the BLS or `ignition.firstboot`. coreos-installer does do this.

I think it'd make sense to teach kola to call `zipl` too. We have code for this in cosa already: https://github.com/coreos/coreos-assembler/blob/03d47c76614c5f54cec83a76b9d1aa22d63653d9/src/gf-platformid#L52-L68. Maybe simplest is to factor that out into a separate script and both kola and gf-platformid can call out to it.

Or maybe better is to flip this around and teach mantle to run `zipl` by extending [the libguestfs code we already have there](https://github.com/coreos/coreos-assembler/blob/03d47c76614c5f54cec83a76b9d1aa22d63653d9/mantle/platform/qemu.go#L649) and then both `gf-platformid` and kola can also call out to it.

**Comment by dustymabe** [comment]
can we create a new issue as an RFE to have the `--kargs` and `--firstbootkargs` work on s390x and then we can point at that issue when disabling these tests on s390x platform.

**Comment by jschintag** [comment]
> can we create a new issue as an RFE to have the `--kargs` and `--firstbootkargs` work on s390x and then we can point at that issue when disabling these tests on s390x platform.

Hmm, i would actually in that case propose to just change the description and the title of this issue, since it is where the discussion is happening and what was already referenced when disabling the tests for s390x. At least i did for the latest one. Might just add an update to that PR for the other issues.

**Comment by dustymabe** [comment]
> > can we create a new issue as an RFE to have the `--kargs` and `--firstbootkargs` work on s390x and then we can point at that issue when disabling these tests on s390x platform.
> 
> Hmm, i would actually in that case propose to just change the description and the title of this issue, since it is where the discussion is happening and what was already referenced when disabling the tests for s390x. At least i did for the latest one. Might just add an update to that PR for the other issues.

That's fine - sometimes new issues are better because they are a bit more focused (they start with the root cause up front), but it's a tricky balance.

Also the RFE would be for COSA IIUC.

Either way works.

**Comment by jschintag** [comment]
>Also the RFE would be for COSA IIUC.

Right, the issue should be for coreos-assembler. Let me create that.

**Comment by jschintag** [comment]
Since the root cause has been determined and the RFE filed, i'll close this issue.
Further discussion should be in https://github.com/coreos/coreos-assembler/issues/2776
